app/javascript/lib/validation/expression-matrices-validation.js

```diff
@@ -0,0 +1,265 @@
+    dataObj.correctNumberOfColumns = dataObj.correctNumberOfColumns ? dataObj.correctNumberOfColumns : line.split(' ').length
+    if (dataObj.correctNumberOfColumns !== line.split(' ').length) {
+      dataObj.rowsWithWrongColumnNumbers.push(lineNum)
+    }
+  }
+
+  const numBadRows = dataObj.rowsWithWrongColumnNumbers.length
+  if (isLastLine && numBadRows > 0) {
+    const rowText = numBadRows > 1 ? 'rows' : 'row'
+    const maxLinesToReport = 10
+    let notedBadRows = dataObj.rowsWithWrongColumnNumbers.slice(0, maxLinesToReport).join(', ')
+    if (numBadRows - maxLinesToReport > 0) {
+      notedBadRows += ` and ${numBadRows - maxLinesToReport} more rows`
+    }
+
+    const msg = `All rows must have the same number of columns - ` +
+      `please ensure the number of columns for ${rowText}: ${notedBadRows}, ` +
+      `matches the specificed number of columns-per-row.`
+    issues.push(['error', 'format:mismatch-column-number', msg])
+  }
+
+  return issues
+}
+
+
+/**
+ * Validate all values are numbers outside first column cell name
+ */
+function validateValuesAreNumeric(line, isLastLine, lineNum, dataObj) {
+  const issues = []
+  dataObj.rowsWithNonNumericValues = dataObj.rowsWithNonNumericValues ? dataObj.rowsWithNonNumericValues : []
+  // skip first column
+  const lineWithoutFirstColumn = line.slice(1)
+
+  if (lineWithoutFirstColumn.some(isNaN)) {
+    dataObj.rowsWithNonNumericValues.push(lineNum)
+  }
+
+  const numBadRows = dataObj.rowsWithNonNumericValues.length
+  if (isLastLine && numBadRows > 0) {
+    const rowText = numBadRows > 1 ? 'rows' : 'row'
+    const maxLinesToReport = 10
+    let notedBadRows = dataObj.rowsWithNonNumericValues.slice(0, maxLinesToReport).join(', ')
+    if (numBadRows - maxLinesToReport > 0) {
+      notedBadRows += ` and ${numBadRows - maxLinesToReport} more rows`
+    }
+
+    const msg = `All values (other than the first column and header row) in a dense matrix file must be numeric. ` +
+      `Please ensure all values in ${rowText}: ${notedBadRows}, are numbers.`
+    issues.push(['error', 'format:invalid-type:not-numeric', msg])
+  }
+
+  return issues
+}
+
+
+/**
+ * Verify "GENE" is present as the first column in the first row for an Expression Matrix file
+ * Todo: Accept files that are R-formatted as well via this ticket: SCP-3971
+ */
+function validateGeneInHeader(headers) {
+  const issues = []
+  if (headers[0].toUpperCase() !== 'GENE') {
+    const msg = 'Dense matrices require the first value of the file to be "GENE". ' +
+       `However, the first value for this file currently is "${headers[0]}".`
+    issues.push(['error', 'format:cap:missing-gene-column', msg])
+  }
+
+  return issues
+}
+
+/**
+   * Validate the first line in the sparse Matrix begins with the string '%%MatrixMarket'
+   */
+function validateMTXHeaderLine(line, lineNum, dataObj) {
+  const issues = []
+  const mtxHeader = line.slice(0, 14)
+  if (lineNum !== 0) {
+    return issues
+  }
```

Comment:
Refining for speed and simplicity, per blocker comment:

```suggestion
/**
* Validate the first line in the sparse matrix begins with the string '%%MatrixMarket'
*/
function validateMTXHeaderLine(line) {
  const issues = []
  const mtxHeader = line.slice(0, 14)
```